clippy_lints/src/misc_early/mixed_case_hex_literals.rs

```diff
@@ -1,32 +1,53 @@
-use clippy_utils::diagnostics::span_lint;
+use clippy_utils::diagnostics::span_lint_and_help;
 use rustc_lint::EarlyContext;
 use rustc_span::Span;
 
 use super::MIXED_CASE_HEX_LITERALS;
 
 pub(super) fn check(cx: &EarlyContext<'_>, lit_span: Span, suffix: &str, lit_snip: &str) {
-    let Some(maybe_last_sep_idx) = lit_snip.len().checked_sub(suffix.len() + 1) else {
-        return; // It's useless so shouldn't lint.
+    let num_end_idx = if lit_snip.ends_with(&format!("_{suffix}")) {
+        lit_snip.len() - (suffix.len() + 1)
+    } else if lit_snip.ends_with(suffix) {
+        lit_snip.len() - suffix.len()
+    } else {
+        lit_snip.len()
     };
```

Comment:
Rather than building a `String` object, can't you test with something like (untested):

```suggestion
    let num_end_idx = match lit_snip.strip_suffix(suffix) {
        Some(p) if p.ends_with('_') => lit_snip.len() - (suffix.len() + 1),
        Some(_) => lit_snip.len() - suffix.len(),
        None => lit_snip.len()
    };
```

Reply:
Makes sense. Thanks for  pointing! https://github.com/rust-lang/rust-clippy/pull/14235/commits/01ee6c41e10de8f5b17dec2163246ad25a847abb